### ISSUE TITLE
NULL pointer check

### ISSUE THREAD

**Issue by tathanhdinh** [ufrisk/MemProcFS#5]
Hello,

Sorry to bother you for such a trivial concern. In reading the code, I've found that sometimes the NULL pointers are not checked, e.g.

https://github.com/ufrisk/MemProcFS/blob/4a1217906c0bc5a637070c7c21fc81c9c2ea0ab6/vmm/vmmvfs.c#L173-L198

where there is no check for `wcsFileName`, so we can pass a NULL pointer via the exported API `VMMDLL_VfsRead`.

IMHO, a NULL pointer dereference in `C` is a [undefined-behavior](https://software.intel.com/en-us/blogs/2015/04/20/null-pointer-dereferencing-causes-undefined-behavior), i.e. if we don't check against NULL, then the behavior of a such a dereference would be unexpected: the compiler can arbitrary select a behavior here and it may not even a crash because of address zero accessing.

So I think that might lead to a minor security issue, but I'm not sure.

Thank you in advance for any response.

**Comment by ufrisk** [comment]
Thanks. It's certainly a code quality issue, one that is relatively easy to fix so I'll look into it in the next update I push.

I however disagree that it's a security issue. The API is only callable from within an application and in the header it's clearly marked with the _In_ SAL definition, not the _In_opt_ or similar so sending NULL in there violates the contract.

But as mentioned above, the fix is easy to implement and code quality improvements are always nice. Thanks.

**Comment by tathanhdinh** [comment]
Thank for the response,

>  it's clearly marked with the `_In_` SAL definition

Ha, I didn't know such a thing exists, thank for notifying that, indeed Microsoft docs says:

> ...annotated pointer parameters are assumed to be required—they must be non-NULL for the function to succeed.

That's great because I'm now can use the SAL to figure out the constraints on arguments.

Thank you.

**Comment by ufrisk** [comment]
:) you may close the issue if you wish. regardless I guess it's not hurting anything if I add the null-checks anyway.

**Comment by tathanhdinh** [comment]
Thank you,

About SAL, should we use `_In_opt_` for `pvReserved` at

https://github.com/ufrisk/MemProcFS/blob/4a1217906c0bc5a637070c7c21fc81c9c2ea0ab6/vmm/vmmdll.h#L155-L159

because it's passed as `NULL` at

https://github.com/ufrisk/MemProcFS/blob/4a1217906c0bc5a637070c7c21fc81c9c2ea0ab6/vmm/vmmdll.h#L164-L165

**Comment by ufrisk** [comment]
Yes, you are correct, pvReserved should be `_In_opt_` - a mistake from my side, which will be changed in next update. It seems like Visual Studio doesn't flag it - probably since that parameter isn't used yet.
